addons/website_blog/controllers/main.py

```diff
@@ -130,6 +163,9 @@ def blog(self, blog=None, tag=None, page=1, **opt):
         blog_url = QueryURL('', ['blog', 'tag'], blog=blog, tag=tag, date_begin=date_begin, date_end=date_end)
 
         blog_posts = BlogPost.search(domain, order="post_date desc")
+
+        first_published_blog_post = (p for p in blog_posts if (p.website_published == True ) ).__next__()
```

Comment:
+1 for __next__

but keep generator ;) it do the same that code of qsm:

```python
>>> def gen_nums():
...     n = 0
...     while n < 4:
...         yield n
...         n += 1
...         print(n)

>>> for n in gen_nums():
...     if n == 2:
...         break
... 
1
2
```

Reply:
Hello @JKE-be , I can't make your example work. I'm trying qsm's one and it seems to work fine..

```python
first_published_blog_post = None
for p in posts:
    if p.website_published:
          first_published_blog_post = p;
          break
```

Reply:
which example... ? just use the public next() method instead of private __next__ 

**Method 1 -> generator called only once**
first_blog = next((p for p in blog_posts if p.website_published))  # take the first one

**Method 2 -> use generator**
blogs = (p for p in blog_posts if p.website_published)
first = next(blogs)  -> return first
second =next(blogs)  -> return second

Reply:
and nooooo... in python we don't use ; at the end of a line ;)

Reply:
```python
published_posts = (p for p in posts if p.website_published)
        if published_posts:
            first_published_blog_post = next(published_posts)
        else:
            first_published_blog_post = Null
```

seems to work fine, thanks

Reply:
*`None`

Reply:
`published_posts = next((p for p in posts if p.website_published), None)`

Reply:
> published_posts = next((p for p in posts if p.website_published), None)

@jke, just to be sure... but is this what qsm wanted to avoid in his first comment (create the whole list for nothing)?
https://github.com/odoo/odoo/pull/34609/commits/12194d4aec2dc0fc878e42308db7a4792d6a95a6#r312478163

Reply:
Ignore me, I am a python noob too :) this does not really create a list

Reply:
@stefanorigano 

yes 

but (i for i in range(1,5)) is an generator iterable
while [i for i in range(1,5)] is a list iterable 

```python
In [11]: (i for i in range(0,5))                                                                                                                                                                                   
Out[11]: <generator object <genexpr> at 0x7fb5e72332a0>

In [12]: [i for i in range(0,5)]                                                                                                                                                                                   
Out[12]: [0, 1, 2, 3, 4]

```

Reply:
Ok... so I'm using this (to handle the cover too)
```python
first_published_blog_post = next((p for p in posts if p.website_published), None)
if first_published_blog_post:
    first_published_blog_post_cover_properties = json.loads(first_published_blog_post.cover_properties)
else:
    first_published_blog_post_cover_properties = None
```

Reply:
```python

first_cover = first_post and json.loads(first_post.cover_properties)
```